ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinIncentivizedAdListener.java

```diff
@@ -0,0 +1,161 @@
+
+    private boolean            mFullyWatched;
+    private AppLovinRewardItem mRewardItem;
+
+    AppLovinIncentivizedAdListener(ApplovinAdapter adapter, MediationRewardedVideoAdListener mediationRewardedVideoAdListener)
+    {
+        mAdapter = adapter;
+        mMediationRewardedVideoAdListener = mediationRewardedVideoAdListener;
+    }
+
+    //
+    // Ad Display Listener
+    //
+
+    @Override
+    public void adDisplayed(AppLovinAd ad)
+    {
+        ApplovinAdapter.log( DEBUG, "Rewarded video displayed" );
+        mMediationRewardedVideoAdListener.onAdOpened( mAdapter );
+    }
+
+    @Override
+    public void adHidden(AppLovinAd ad)
+    {
+        ApplovinAdapter.log( DEBUG, "Rewarded video dismissed" );
+
+        if ( mFullyWatched && mRewardItem != null )
+        {
+            mMediationRewardedVideoAdListener.onRewarded( mAdapter, mRewardItem );
+        }
+
+        mMediationRewardedVideoAdListener.onAdClosed( mAdapter );
+
+        // Clear states in the case this listener gets re-used in the future
+        mFullyWatched = false;
+        mRewardItem = null;
+    }
+
+    //
+    // Ad Click Listener
+    //
+
+    @Override
+    public void adClicked(AppLovinAd ad)
+    {
+        ApplovinAdapter.log( DEBUG, "Rewarded video clicked" );
+
+        mMediationRewardedVideoAdListener.onAdClicked( mAdapter );
+        mMediationRewardedVideoAdListener.onAdLeftApplication( mAdapter );
+    }
+
+    //
+    // Video Playback Listener
+    //
+
+    @Override
+    public void videoPlaybackBegan(AppLovinAd ad)
+    {
+        ApplovinAdapter.log( DEBUG, "Rewarded video playback began" );
+        mMediationRewardedVideoAdListener.onVideoStarted( mAdapter );
+    }
+
+    @Override
+    public void videoPlaybackEnded(AppLovinAd ad, double percentViewed, boolean fullyWatched)
+    {
+        ApplovinAdapter.log( DEBUG, "Rewarded video playback ended at playback percent: " + percentViewed + "%" );
+        mFullyWatched = fullyWatched;
+    }
+
+    //
+    // Reward Listener
+    //
+
+    @Override
+    public void userOverQuota(AppLovinAd ad, Map response)
+    {
+        ApplovinAdapter.log( ERROR, "Rewarded video validation request for ad did exceed quota with response: " + response );
+    }
+
+    @Override
+    public void validationRequestFailed(AppLovinAd ad, int code)
+    {
+        ApplovinAdapter.log( ERROR, "Rewarded video validation request for ad failed with error code: " + code );
+    }
+
+    @Override
+    public void userRewardRejected(AppLovinAd ad, Map response)
+    {
+        ApplovinAdapter.log( ERROR, "Rewarded video validation request was rejected with response: " + response );
+    }
+
+    @Override
+    public void userDeclinedToViewAd(AppLovinAd ad)
+    {
+        ApplovinAdapter.log( DEBUG, "User declined to view rewarded video" );
+    }
+
+    @Override
+    public void userRewardVerified(AppLovinAd ad, Map response)
```

Comment:
We can use stronger typing with `Map<String, String>`

Reply:
Good call, I thought I tried that already but didn't compile 😅